src/renderers/webgl/WebGLMaterials.js

```diff
@@ -0,0 +1,677 @@
+
+		uniforms.diffuse.value.copy( material.color );
+
+	}
+
+	if ( material.emissive ) {
+
+		uniforms.emissive.value.copy( material.emissive ).multiplyScalar( material.emissiveIntensity );
+
+	}
+
+	if ( material.map ) {
+
+		uniforms.map.value = material.map;
+
+	}
+
+	if ( material.alphaMap ) {
+
+		uniforms.alphaMap.value = material.alphaMap;
+
+	}
+
+	if ( material.specularMap ) {
+
+		uniforms.specularMap.value = material.specularMap;
+
+	}
+
+	var envMap = material.envMap || environment;
+
+	if ( envMap ) {
+
+		uniforms.envMap.value = envMap;
+
+		uniforms.flipEnvMap.value = envMap.isCubeTexture ? - 1 : 1;
+
+		uniforms.reflectivity.value = material.reflectivity;
+		uniforms.refractionRatio.value = material.refractionRatio;
+
+		uniforms.maxMipLevel.value = _properties.get( envMap ).__maxMipLevel;
+
+	}
+
+	if ( material.lightMap ) {
+
+		uniforms.lightMap.value = material.lightMap;
+		uniforms.lightMapIntensity.value = material.lightMapIntensity;
+
+	}
+
+	if ( material.aoMap ) {
+
+		uniforms.aoMap.value = material.aoMap;
+		uniforms.aoMapIntensity.value = material.aoMapIntensity;
+
+	}
+
+	// uv repeat and offset setting priorities
+	// 1. color map
+	// 2. specular map
+	// 3. normal map
+	// 4. bump map
+	// 5. alpha map
+	// 6. emissive map
+
+	var uvScaleMap;
+
+	if ( material.map ) {
+
+		uvScaleMap = material.map;
+
+	} else if ( material.specularMap ) {
+
+		uvScaleMap = material.specularMap;
+
+	} else if ( material.displacementMap ) {
+
+		uvScaleMap = material.displacementMap;
+
+	} else if ( material.normalMap ) {
+
+		uvScaleMap = material.normalMap;
+
+	} else if ( material.bumpMap ) {
+
+		uvScaleMap = material.bumpMap;
+
+	} else if ( material.roughnessMap ) {
+
+		uvScaleMap = material.roughnessMap;
+
+	} else if ( material.metalnessMap ) {
+
+		uvScaleMap = material.metalnessMap;
+
+	} else if ( material.alphaMap ) {
+
+		uvScaleMap = material.alphaMap;
+
+	} else if ( material.emissiveMap ) {
+
+		uvScaleMap = material.emissiveMap;
+
+	}
+
+	if ( uvScaleMap !== undefined ) {
+
+		// backwards compatibility
+		if ( uvScaleMap.isWebGLRenderTarget ) {
+
+			uvScaleMap = uvScaleMap.texture;
+
+		}
+
+		if ( uvScaleMap.matrixAutoUpdate === true ) {
+
+			uvScaleMap.updateMatrix();
+
+		}
+
+		uniforms.uvTransform.value.copy( uvScaleMap.matrix );
+
+	}
+
+	// uv repeat and offset setting priorities for uv2
+	// 1. ao map
+	// 2. light map
+
+	var uv2ScaleMap;
+
+	if ( material.aoMap ) {
+
+		uv2ScaleMap = material.aoMap;
+
+	} else if ( material.lightMap ) {
+
+		uv2ScaleMap = material.lightMap;
+
+	}
+
+	if ( uv2ScaleMap !== undefined ) {
+
+		// backwards compatibility
+		if ( uv2ScaleMap.isWebGLRenderTarget ) {
+
+			uv2ScaleMap = uv2ScaleMap.texture;
+
+		}
+
+		if ( uv2ScaleMap.matrixAutoUpdate === true ) {
+
+			uv2ScaleMap.updateMatrix();
+
+		}
+
+		uniforms.uv2Transform.value.copy( uv2ScaleMap.matrix );
+
+	}
+
+};
+
+WebGLMaterials.prototype.refreshUniformsLine = function ( uniforms, material ) {
```

Comment:
Please organize the code similar to other WebGL classes like:

https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLProperties.js

This will automatically remove all internal functions from the public interface.

Reply:
done, "sealed" all private functions, extending forbidden

Reply:
I don't think having `_properties` in module scope is save. If you create two renderers, you will overwrite this value.

> done, "sealed" all private functions, extending forbidden

The idea is to move this functions into the ctor of `WebGLMaterials`. The public interface is exported like so:
```js
return {
	refreshUniforms: refreshUniforms,
	refreshUniformsFog: refreshUniformsFog
};
```

Reply:
> I don't think having `_properties` in module scope is save

agree, my bad

> The idea is to move the code into the ctor of WebGLMaterials.

done

Reply:
Looks much better now.